### PR TITLE
GitHub actions

### LINKED ISSUE #8: Migrate current Travis pipeline to GitHub action
## Description
Migrate to legacy TravisCI deployment logic to GitHub actions

## Acceptance Criteria
- [x] Artefact deployed to charts-ose.clearmatics.com AWS picket
- [x] PR logic is maintained
- [x] Tag logic is maintained
- [x] Tag permissions logic is maintained

## Note
Unfortunately there is a known problem and an [open issue](https://github.com/helm/helm/issues/5861) with deploying helm charts using Helm 3 to Docker Hub as originally planned. Have restricted the scope of this ticket to push new release to the existing charts location listed above in the Acceptance Criteria

### PR DESCRIPTION
**Pull request by Klazomenai**
### Changelog
Resolves #8 

**Note:** Permissions logic on who can perform Pull Request and Tagging has been moved out of the CI/CD level and to the permissions on the repository level.

### All Submissions
* [x] Have you followed the guidelines in our [Contributing document](CONTRIBUTING.md)?

### CODE REVIEW

File: .github/workflows/pr.yaml [hunk 1]
```diff
@@ -0,0 +1,56 @@
+name: "Integration Tests"
+
+on:
+  pull_request:
+    branches: [ master ]
+
+jobs:
+  integration_tests:
+    runs-on: ubuntu-latest
```
Comment (eastata) [inline .github/workflows/pr.yaml]: Could we use latest `ubuntu-20.04` instead?

File: .github/workflows/publish-artefact.yaml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+on:
+  push:
+    tags:
+    - 'autonity-v[0-9]+.[0-9]+.[0-9]+*'
+    - 'audonity-demo-v[0-9]+.[0-9]+.[0-9]+*'
+    - 'autonity-network-v[0-9]+.[0-9]+.[0-9]+*'
+
+jobs:
+  integration_tests_publish_on_tag:
+    runs-on: ubuntu-latest
```
Comment (eastata) [inline .github/workflows/publish-artefact.yaml]: Could we use latest `ubuntu-20.04` instead?

File: .github/workflows/pr.yaml [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    - name: Setup gcloud environment
+      uses: GoogleCloudPlatform/github-actions/setup-gcloud@0.1.2
+      with:
+        version: '290.0.1'
+        project_id: ${{ secrets.GCP_PROJECT_ID }}
+        service_account_key: ${{ secrets.GCP_SA_KEY }}
+        export_default_credentials: true
+
+    - name: Configure the kubeconfig
+      run: gcloud container clusters get-credentials tf-gke-helm --zone ${{ secrets.GCP_PROJECT_ZONE }} --project ${{ secrets.GCP_PROJECT_ID }}
```
Comment (eastata) [inline .github/workflows/pr.yaml]: `tf-gke-helm` ==> `${{secrets.GCP_PROJECT_ID}}`

File: .github/workflows/publish-artefact.yaml [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    - name: Setup gcloud environment
+      uses: GoogleCloudPlatform/github-actions/setup-gcloud@0.1.2
+      with:
+        version: '290.0.1'
+        project_id: ${{ secrets.GCP_PROJECT_ID }}
+        service_account_key: ${{ secrets.GCP_SA_KEY }}
+        export_default_credentials: true
+
+    - name: Configure the kubeconfig
+      run: gcloud container clusters get-credentials tf-gke-helm --zone ${{ secrets.GCP_PROJECT_ZONE }} --project ${{ secrets.GCP_PROJECT_ID }}
```
Comment (eastata) [inline .github/workflows/publish-artefact.yaml]: `tf-gke-helm` ==> `${{secrets.GCP_PROJECT_ID}}`

File: .github/workflows/publish-artefact.yaml [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+    - name: Install helm
+      uses: azure/setup-helm@v1
+      with:
+        version: '3.2.4'
+      id: install
+
+    - name: Deploy the chart
+      run: |
+        helm install $GITHUB_HEAD_REF-$SHORT_SHA $CHART_PATH --namespace $GITHUB_ACTOR
```
Comment (eastata) [inline .github/workflows/publish-artefact.yaml:49]: Here we have to check if chart `$CHART_PATH` exist and matched with `git tag` to have a guaranty that we release correct chart/version

File: .github/workflows/publish-artefact.yaml [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+    - name: Publish the chart
+      run: |
+        aws s3 cp ./repo/${GITHUB_REF##*/}.tgz s3://${S3_BUCKET_NAME}
+
+    - name: Publish the main index
+      run: |
+        aws s3 cp ./repo/index.yaml s3://${S3_BUCKET_NAME} --metadata-directive REPLACE --cache-control max-age=0,no-cache,no-store,must-revalidate --content-type "text/plain"
+    - name: Cleanup
+      run: helm delete $GITHUB_HEAD_REF-$SHORT_SHA
```
Comment (eastata) [inline .github/workflows/publish-artefact.yaml:81]: Add:
* kubectl delete namespace
* cleanup even if build/tests failed

File: .github/workflows/publish-artefact.yaml [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    - name: Deploy the chart
+      run: |
+        helm install $GITHUB_HEAD_REF-$SHORT_SHA $CHART_PATH --namespace $GITHUB_ACTOR
+        # Sadly --wait fails, so have to sleep, look to maybe pull in mock test to solve
+        # this problem another way
+        sleep 60
+    - name: Run the tests
+      run: helm test $GITHUB_HEAD_REF-$SHORT_SHA
+
+    - name: Configure AWS credentials from Test account
```
Comment (eastata) [inline .github/workflows/publish-artefact.yaml:56]: Is any reason why we are still using `aws` instead `google` or `DockerHub` registry?

**Review by eastata — CHANGES_REQUESTED** [review]

**Comment by Klazomenai** [comment]
Regarding CI failures, there is an exiting ticket to address this: https://github.com/clearmatics/charts-ose-helm3/issues/10

This issue does not need to be resolved now and is being accepted as tech debt.

File: .github/workflows/publish-artefact.yaml (hunk 6 above)
Comment (Klazomenai) [reply .github/workflows/publish-artefact.yaml:81]: No longer an issue with Helm 3

File: .github/workflows/publish-artefact.yaml (hunk 7 above)
Comment (Klazomenai) [reply .github/workflows/publish-artefact.yaml:56]: See issue linked from description as to why not using DockerHub. Using `aws` as that is the current working standard.

File: .github/workflows/pr.yaml (hunk 3 above)
Comment (Klazomenai) [reply .github/workflows/pr.yaml]: Done

File: .github/workflows/publish-artefact.yaml (hunk 4 above)
Comment (Klazomenai) [reply .github/workflows/publish-artefact.yaml]: Done

File: .github/workflows/pr.yaml (hunk 1 above)
Comment (Klazomenai) [reply .github/workflows/pr.yaml]: Done

File: .github/workflows/publish-artefact.yaml (hunk 2 above)
Comment (Klazomenai) [reply .github/workflows/publish-artefact.yaml]: Done

File: .github/workflows/publish-artefact.yaml (hunk 5 above)
Comment (Klazomenai) [reply .github/workflows/publish-artefact.yaml:49]: Will be resolved here: https://github.com/clearmatics/charts-ose-helm3/issues/10

**Review by eastata — APPROVED** [review]

**Pull request merged**